README.md
# BeamOlympics

![Beam Olympics](https://s3.amazonaws.com/uploads.hipchat.com/15025/55900/LeqcbfiQzCN48Lb/Beam%20Olympics%20black.png)

Let's find the fastest beamer!

## Introduction
**BeamOlympics** is an Erlang app to help you check your Erlang-Fu. You can also play with friends in a local network.

## Compilation
If you want to generate a release by hand, follow the steps below. If you would like to start with a pre-generated release, just download [the latest one](https://github.com/inaka/beam_olympics/releases) and jump to the [Start](#start) section below.

### Prerequisites
As stated in rebar.config, **BeamOlympics** server only compiles with Erlang/OTP19+.

### Installation
Before you can start playing, you have to install the **BeamOlympics** server.
To do so, clone this repo and use [reba3](https://github.com/erlang/rebar3) to generate a release for it.

```bash
$ git clone https://github.com/inaka/beam_olympics.git
$ rebar3 release
```

If needed, between those 2 steps you can change your node name in [vm.args](https://github.com/inaka/beam_olympics/blob/master/config/vm.args#L2) to your convenience.

## Start
To start the **BeamOlympics** server you you run…

```bash
$ $REL_PATH/beam_olympics/bin/beam_olympics start
```

Whre `$REL_PATH` would be `_build/default/rel` if you just generated the release with rebar3 or `.` if you downloaded a pre-compiled release from github.

That will boot up an Erlang node in your computer with the name specified by vm.args.

## Players
Inside the node started in the previous step, a [gen_server](http://erldocs.com/maint/stdlib/gen_server.html) will be running. That server will act as the point of contact for clients to play the game.
Every interaction will be accomplished through `gen_server` calls. For example, to check the overall game statistics, from a client node a caller can evaluate the following Elixir code:

```elixir
GenServer.call({:bo_server, server_node}, :stats)
```

where `server_node` is a variable bound to the server node name (or a function that returns it).

### General Rules
* To prevent fraud, users must always post solutions from the node in which they signed up. The node might crash/be closed, but when restarted it has to be called with the same name, otherwise submissions will be rejected.
* There is no *undo*, if you skipped a task you can't choose to submit a solution for it later.
* Each task has it's own associated score. Points are assigned this way:
    - Each time a task is correctly solved, `score` points are added to the user
    - Each time a task is skipped, `score / 2` points are deduced from the user
* After skipping or completing a task, the user will either receive a new task or just the atom `the_end` indicating that the game is over

### Commands
You can find the complete list of commands/calls with their possible answers below. Specs are written as functions although in reality the input is the call sent to the server and the output is the possible responses the server may return.

----
#### Signup
Using this call, the users register themselves in the game and receive their initial task.

##### Erlang Specs
```erlang
{signup, PlayerName :: binary()} -> {ok, Task} | {error, conflict}.
Task :: #{ name := module()
         , desc := binary()
         , spec := Spec
         , score := pos_integer()
         }.
Spec :: #{ input := [binary()]
         , output := binary()
         }.
```

##### Elixir Specs
```elixir
{:signup, player_name} :: {:ok, task} | {:error, :conflict}
  @type player_name :: String.t
  @type task :: %{
                  name: module,
                  desc: String.t,
                  spec: spec,
                  score: pos_integer
                 }
  @type spec :: %{
                  input: [String.t],
                  output: String.t
                 }
```

##### Details
* The server will return `conflict` if the player is already registered.
* The spec describes the function required to complete the task. `input` indicates the number and types of the parameters, `output` indicates the type of result it should produce.

##### Examples
```erlang
gen_server:call({bo_server, ServerNode}, {signup, <<"Player">>}).
```

```elixir
GenServer.call({:bo_server, server_node}, {:signup, "Player"})
```

----
#### Task
Retrieves the current task for the player.

##### Erlang Specs
```erlang
{task, PlayerName :: binary()} ->
    {ok, Task} | {error, ended | forbidden | notfound}
```

##### Elixir Specs
```elixir
{:task, player_name} :: {:ok, task} | {:error, :ended | :forbidden | :notfound}
```

##### Details
* The server will return `ended` if the game is over for the player and therefore there is no current task
* The server will return `forbidden` if the player is calling from a node that's not the one from which they signed up in the first place
* The server will return `notfound` if the player doesn't exist

##### Examples
```erlang
gen_server:call({bo_server, ServerNode}, {task, <<"Player">>}).
```

```elixir
GenServer.call({:bo_server, server_node}, {:task, "Player"})
```

----
#### Submit
Submits a solution for the current task.

##### Erlang Specs
```erlang
{submit, PlayerName :: binary(), Solution :: any()} ->
    {ok, Task}
  | the_end
  | {error, invalid | timeout | ended | forbidden | notfound}
  | {failures, [term(), ...]}
```

##### Elixir Specs
```elixir
{:submit, player_name, term} ::
    {:ok, task} | :the_end |
    {:error, :invalid | :timeout | :ended | :forbidden | :notfound} |
    {:failures, [term,...]}
```

##### Details
* The server will return `ended` if the game is over for the player and therefore there is no current task
* The server will return `forbidden` if the player is calling from a node that's not the one from which they signed up in the first place
* The server will return `notfound` if the player doesn't exist
* The server will return `timeout` if the evaluation of the tests took longer than expected
* The server will return `invalid` if the arity of the provided function doesn't match the arity expected by the task or if the input is not even a function
* The server will return `the_end` if the game is over for the player (i.e. there are no more tasks to complete/skip)
* The server will return `failures` with a list of failure descriptions if there are tests for which the provided function fails

##### Examples
```erlang
gen_server:call(
    {bo_server, ServerNode}, {submit, <<"Player">>, fun() -> something end}).
```

```elixir
GenServer.call(
    {:bo_server, server_node}, {:submit, "Player", fn() -> :something end})
```

----
#### Skip
Skips the current task.

##### Erlang Specs
```erlang
{skip, PlayerName :: binary()} ->
    {ok, Task} | the_end | {error, ended | forbidden | notfound}
```

##### Elixir Specs
```elixir
{:skip, player_name} ::
    {:ok, task} | :the_end | {:error, :ended | :forbidden | :notfound}
```

##### Details
* The server will return `ended` if the game is over for the player and therefore there is no current task
* The server will return `forbidden` if the player is calling from a node that's not the one from which they signed up in the first place
* The server will return `notfound` if the player doesn't exist
* The server will return `the_end` if the game is over for the player (i.e. there are no more tasks to complete/skip)

##### Examples
```erlang
gen_server:call({bo_server, ServerNode}, {skip, <<"Player">>}).
```

```elixir
GenServer.call({:bo_server, server_node}, {:skip, "Player"})
```

----
#### Score
Retrieves the current score for the player.

##### Erlang Specs
```erlang
{score, PlayerName :: binary()} -> {ok, integer()} | {error, forbidden | notfound}
```

##### Elixir Specs
```elixir
{:score, player_name} :: {:ok, integer} | {:error, :forbidden | :notfound}
```

##### Details
* The server will return `forbidden` if the player is calling from a node that's not the one from which they signed up in the first place
* The server will return `notfound` if the player doesn't exist

##### Examples
```erlang
gen_server:call({bo_server, ServerNode}, {score, <<"Player">>}).
```

```elixir
GenServer.call({:bo_server, server_node}, {:score, "Player"})
```

----
#### Stats
Retrieves the current game stats.

##### Erlang Specs
```erlang
stats -> #{ tasks := pos_integer()
          , players := [PlayerStats]
          }.
PlayerStats :: #{ name := binary()
                , done := non_neg_integer()
                , score := integer()
                }.
```

##### Elixir Specs
```elixir
:stats :: stats
  @type stats :: %{
                    tasks: pos_integer,
                    players: [player_stats]
                  }
  @type player_stats :: %{
                           name: String.t,
                           done: non_neg_integer,
                           score: integer
                         }
```

##### Details
* `tasks` will be the total number of tasks in the game
* `done` will be the number of tasks already completed/skipped by the player

##### Examples
```erlang
gen_server:call({bo_server, ServerNode}, stats).
```

```elixir
GenServer.call({:bo_server, server_node}, :stats)
```

---

## Contact Us
For **questions** or **general comments** regarding the use of this library,
please use our public [hipchat room](http://inaka.net/hipchat).

If you find any **bugs** or have a **problem** while using this library, please
[open an issue](https://github.com/inaka/elvis/issues/new) in this repo
(or a pull request :)).

And you can check all of our open-source projects at [inaka.github.io](http://inaka.github.io).


CHANGELOG.md
# Change Log

## [1.2.3](https://github.com/inaka/beam_olympics/tree/1.2.3) (2016-10-14)
[Full Changelog](https://github.com/inaka/beam_olympics/compare/1.2.2...1.2.3)

**Closed issues:**

- upgrade sumo\_db dependecy [\#55](https://github.com/inaka/beam_olympics/issues/55)

**Merged pull requests:**

- \[\#55\] upgrading sumo dep to 0.7.1 [\#56](https://github.com/inaka/beam_olympics/pull/56) ([ferigis](https://github.com/ferigis))

## [1.2.2](https://github.com/inaka/beam_olympics/tree/1.2.2) (2016-10-12)
[Full Changelog](https://github.com/inaka/beam_olympics/compare/1.2.1...1.2.2)

**Closed issues:**

- Version Bump to 1.2.2 [\#53](https://github.com/inaka/beam_olympics/issues/53)
- Fix sumo:find/2 calls [\#51](https://github.com/inaka/beam_olympics/issues/51)

**Merged pull requests:**

- \[\#53\] version bump to 1.2.2 [\#54](https://github.com/inaka/beam_olympics/pull/54) ([ferigis](https://github.com/ferigis))
- \[\#51\] fix calls to sumo:find/2 [\#52](https://github.com/inaka/beam_olympics/pull/52) ([ferigis](https://github.com/ferigis))

## [1.2.1](https://github.com/inaka/beam_olympics/tree/1.2.1) (2016-10-12)
[Full Changelog](https://github.com/inaka/beam_olympics/compare/1.2.0...1.2.1)

**Closed issues:**

- Version Bump to 1.2.1 [\#49](https://github.com/inaka/beam_olympics/issues/49)
- upgrade sumo\_db dependecy [\#47](https://github.com/inaka/beam_olympics/issues/47)

**Merged pull requests:**

- Version bump to 1.2.1 [\#50](https://github.com/inaka/beam_olympics/pull/50) ([ferigis](https://github.com/ferigis))
- \[\#47\] upgrade sumo\_db to 0.7.0 [\#48](https://github.com/inaka/beam_olympics/pull/48) ([ferigis](https://github.com/ferigis))

## [1.2.0](https://github.com/inaka/beam_olympics/tree/1.2.0) (2016-09-01)
[Full Changelog](https://github.com/inaka/beam_olympics/compare/1.1.1...1.2.0)

**Merged pull requests:**

- Version Bump to 1.2.0 [\#46](https://github.com/inaka/beam_olympics/pull/46) ([elbrujohalcon](https://github.com/elbrujohalcon))
- Added new hooks [\#45](https://github.com/inaka/beam_olympics/pull/45) ([HernanRivasAcosta](https://github.com/HernanRivasAcosta))

## [1.1.1](https://github.com/inaka/beam_olympics/tree/1.1.1) (2016-08-26)
[Full Changelog](https://github.com/inaka/beam_olympics/compare/1.1.0...1.1.1)

**Closed issues:**

- Version Bump to 1.1.1 [\#43](https://github.com/inaka/beam_olympics/issues/43)
- Move from erlang.mk to rebar3 [\#40](https://github.com/inaka/beam_olympics/issues/40)

**Merged pull requests:**

- \[Close \#43\] version bump to 1.1.1 [\#44](https://github.com/inaka/beam_olympics/pull/44) ([Euen](https://github.com/Euen))
- \[Close \#40\] given support to publish in hex.pm and sumo\_db 0.6.1 [\#42](https://github.com/inaka/beam_olympics/pull/42) ([Euen](https://github.com/Euen))

## [1.1.0](https://github.com/inaka/beam_olympics/tree/1.1.0) (2016-08-25)
[Full Changelog](https://github.com/inaka/beam_olympics/compare/1.0.2...1.1.0)

**Merged pull requests:**

- Version bump [\#41](https://github.com/inaka/beam_olympics/pull/41) ([HernanRivasAcosta](https://github.com/HernanRivasAcosta))
- Added the events system [\#39](https://github.com/inaka/beam_olympics/pull/39) ([HernanRivasAcosta](https://github.com/HernanRivasAcosta))

## [1.0.2](https://github.com/inaka/beam_olympics/tree/1.0.2) (2016-07-13)
[Full Changelog](https://github.com/inaka/beam_olympics/compare/1.0.0...1.0.2)

**Fixed bugs:**

- Invalid node on Henchmen test [\#31](https://github.com/inaka/beam_olympics/pull/31) ([elbrujohalcon](https://github.com/elbrujohalcon))
- Invalid user input goes through [\#30](https://github.com/inaka/beam_olympics/pull/30) ([elbrujohalcon](https://github.com/elbrujohalcon))

**Closed issues:**

- Version Bump to 1.0.2 [\#35](https://github.com/inaka/beam_olympics/issues/35)
- Remove all private tasks [\#34](https://github.com/inaka/beam_olympics/issues/34)

**Merged pull requests:**

- Version bump [\#38](https://github.com/inaka/beam_olympics/pull/38) ([HernanRivasAcosta](https://github.com/HernanRivasAcosta))
- Fixed some of the tasks [\#37](https://github.com/inaka/beam_olympics/pull/37) ([HernanRivasAcosta](https://github.com/HernanRivasAcosta))
- Fixed some of the tasks [\#36](https://github.com/inaka/beam_olympics/pull/36) ([HernanRivasAcosta](https://github.com/HernanRivasAcosta))
- Version Bump to 1.0.1 [\#32](https://github.com/inaka/beam_olympics/pull/32) ([elbrujohalcon](https://github.com/elbrujohalcon))

## [1.0.0](https://github.com/inaka/beam_olympics/tree/1.0.0) (2016-07-12)
**Closed issues:**

- Get Current Task [\#11](https://github.com/inaka/beam_olympics/issues/11)
- Basic game logic [\#5](https://github.com/inaka/beam_olympics/issues/5)
- Make the Repo Public [\#24](https://github.com/inaka/beam_olympics/issues/24)

**Merged pull requests:**

- README pieces for Open-Source checklist [\#29](https://github.com/inaka/beam_olympics/pull/29) ([elbrujohalcon](https://github.com/elbrujohalcon))
- Added 4 new tests [\#28](https://github.com/inaka/beam_olympics/pull/28) ([HernanRivasAcosta](https://github.com/HernanRivasAcosta))
- Awesomeness [\#27](https://github.com/inaka/beam_olympics/pull/27) ([elbrujohalcon](https://github.com/elbrujohalcon))
- \[\#problems\] Add henchmen task [\#23](https://github.com/inaka/beam_olympics/pull/23) ([elbrujohalcon](https://github.com/elbrujohalcon))
- \[\#problems\] Add romans problem [\#22](https://github.com/inaka/beam_olympics/pull/22) ([elbrujohalcon](https://github.com/elbrujohalcon))
- Documentation / Examples [\#20](https://github.com/inaka/beam_olympics/pull/20) ([elbrujohalcon](https://github.com/elbrujohalcon))
- Specify OTP19+ in rebar.config [\#19](https://github.com/inaka/beam_olympics/pull/19) ([elbrujohalcon](https://github.com/elbrujohalcon))
- Run tests in parallel [\#18](https://github.com/inaka/beam_olympics/pull/18) ([elbrujohalcon](https://github.com/elbrujohalcon))

src/tasks/bo_romans.erl
-module(bo_romans).

-behaviour(bo_task).

-export([description/0, spec/0, score/0, timeout/0, tests/0]).

-spec description() -> binary().
description() ->
  <<"Roman Numerals: Parse Roman Numerals as binaries and convert them into"
    " integers.\n"
    "Provided function is expected to fail on invalid input."
    >>.

-spec spec() -> bo_task:spec().
spec() -> #{input => [<<"binary()">>], output => <<"pos_integer()">>}.

-spec score() -> 150.
score() -> 150.

-spec timeout() -> 5000.
timeout() -> 5000.

-spec tests() -> [bo_task:test()].
tests() -> [build_test(Input) || Input <- test_inputs()].

test_inputs() ->
  [ {<<"">>,              fail}
  , {<<"WAT">>,           fail}
  , {<<"I">>,             1}
  , {<<"II">>,            2}
  , {<<"III">>,           3}
  , {<<"IV">>,            4}
  , {<<"V">>,             5}
  , {<<"IX">>,            9}
  , {<<"X">>,             10}
  , {<<"XV">>,            15}
  , {<<"XXX">>,           30}
  , {<<"XL">>,            40}
  , {<<"L">>,             50}
  , {<<"XC">>,            90}
  , {<<"C">>,             100}
  , {<<"CC">>,            200}
  , {<<"D">>,             500}
  , {<<"M">>,             1000}
  , {<<"XIV">>,           14}
  , {<<"XIX">>,           19}
  , {<<"CMXLIX">>,        949}
  , {<<"MMMDCCCLXXVI">>,  3876}
  % , {<<"IIII">>,          fail}
  % , {<<"VV">>,            fail}
  % , {<<"IL">>,            fail}
  % , {<<"XM">>,            fail}
  ].

build_test({Roman, fail}) ->
  fun(Fun) ->
    try Fun(Roman) of
      Output -> {error, #{input => Roman, output => Output, expected => fail}}
    catch
      _:_ -> ok
    end
  end;
build_test({Roman, Decimal}) ->
  fun(Fun) ->
    try Fun(Roman) of
      Decimal -> ok;
      Output ->
        {error, #{input => Roman, output => Output, expected => Decimal}}
    catch
      _:Error ->
        {error, #{input => Roman, output => Error, expected => Decimal}}
    end
  end.
